_posts/2018-05-23-getting-foreman-search-results-into-your-puppet-manifest.md

```diff
@@ -0,0 +1,101 @@
+---
+layout: post
+title: update - Getting foreman search results into your Puppet manifest
+date: 2018-05-23 11:19:42
+author: Fabien Combernous
+tags:
+- foreman
+- puppet
+- puppetdb
+- monitoring
+- backup
+- identity management
+---
+
+How to use the foreman search language inside puppet manifests and knowing informations about others nodes.
+This page is an update of the [previous post written by Ohad Levy](https://theforeman.org/2012/01/getting-foreman-search-results-into.html)
+
+<!--more-->
+
+## Why ?
+
+When you start puppet, in general you start simply. And so innocently, each node is more or less alone in the world. And puppet, in this way, can already help you so much.
+
+But fastly, you understand that in some situations, it is interesting to know informations from others nodes. And this happend tipically about monitoring, backups or identity management. For instance, you add a new node into your puppet infrastructure. And your monitoring server takes into account this new server by adding the configuration relating to its monitoring. 
+
+In a standard way, in the puppet galaxy it exists the option `storeconfigs` of pupper server and linked [puppetdb server](https://puppet.com/docs/puppetdb/latest). A puppetdb server records every node's catalogs, facts and reports. And the immediate usage is exported resources (a double “at” sign `@@` in manifests). 
+
+But, if you don't have a puppetdb server and you are already using theforeman as external node classifier (aka ENC), you can also make fun stuffs.
+
+## A use case out of standard usages.
+
+Use cases about monitoring or backups are standards. Lets try to build an other example where we can use the foreman search engine.
+
+Imagine that you are managing nodes with puppet and they are in `hostgroups` like `secure_base/role`. Where `secure_base` permits to apply all common settings to all your nodes in a secured environment. And `role` permits to apply settings dedicated to the role of a node. Now, lets say you have the constraint to not use a dns. And your customer would like to call hosts in a humain way, so with names and not with IPs. I already met this use case. Fortunatly, servers were using only one ip address.
+
+So you have to manage your `/etc/hosts` with puppet or you become crazy. 
+
+If each node execute the following code :
+
+```
+  $query = foreman({foreman_user => 'apiuser',
+                    foreman_pass => 'here_a_password',
+                    item         => 'hosts',
+                    search       => 'hostgroup_fullname ~ secure_base',
+  })
+
+  $nodes = $query['results']
+
+```
+
+In `$nodes`, you got all the needed informations about others nodes. You can check my statement by adding the following peace of code :
```

Comment:
`informations about others nodes` -> `information about other nodes`
`peace of code` -> `piece of code`